src/main/java/org/assertj/core/util/introspection/Introspection.java

```diff
@@ -102,19 +108,47 @@ private static boolean isValidGetter(Method method) {
   }
 
   private static Method findMethod(String name, Object target) {
-    Class<?> clazz = target.getClass();
+    final MethodKey methodKey = new MethodKey(name, target.getClass());
+    return METHOD_CACHE.computeIfAbsent(methodKey, Introspection::findMethodByKey).orElse(null);
```

Comment:
I believe the code can be simplified if `findMethodByKey` returned `Method`, you wouldn't need `.orElse(null)`.
Optional are useful to force client to deal with the absence of value but since we are returning in this method `findMethodByKey` could also do that.

Reply:
So the reason I went for Optional is to cache negative results. Otherwise in our case - bare getters - every subsequent lookup for 'getXYZ' executes the search logic all over again, with all exceptions being thrown etc
Since ConcurrentHashMap does not allow null values, I needed something to represent missing value - empty optional looked like the most obvious choice.
I certainly don't want to compromise thread-safety by using regular HashMap
It's possible to use some sentinel value (Object::toString?) - that would actually slightly reduce memory footprint, but not necessarily improve the code
It's possible to inspect the class as a whole and cache some richer object similar to BeanInfo per class (and that would potentially benefit field reflection as well) - that's a bigger change, but I'm prepared to give it a go
Happy to hear other ideas

Reply:
that's alright, leave the optional, I did not remember concurrent hash map required non null values, add a comment for that for me when I and my gold fish memory read the code again in 3 months. :D